controllers/gitrepository_controller.go

```diff
@@ -87,410 +84,505 @@ func (r *GitRepositoryReconciler) SetupWithManagerAndOptions(mgr ctrl.Manager, o
+		obj.Status.Artifact = nil
+		obj.Status.URL = ""
+	}
+
+	// Record that we do not have an artifact
+	if obj.GetArtifact() == nil {
+		conditions.MarkTrue(obj, sourcev1.ArtifactUnavailableCondition, "NoArtifact", "No artifact for resource in storage")
+		return ctrl.Result{Requeue: true}, nil
+	}
+	conditions.Delete(obj, sourcev1.ArtifactUnavailableCondition)
+
+	// Always update URLs to ensure hostname is up-to-date
+	// TODO(hidde): we may want to send out an event only if we notice the URL has changed
+	r.Storage.SetArtifactURL(obj.GetArtifact())
+	obj.Status.URL = r.Storage.SetHostname(obj.Status.URL)
+
+	return ctrl.Result{RequeueAfter: obj.Spec.Interval.Duration}, nil
+}
 
-	// determine auth method
+// reconcileSource ensures the upstream Git repository can be reached and checked out using the declared configuration,
+// and observes its state.
+//
+// The repository is checked out to the given dir using the defined configuration, and in case of an error during the
+// checkout process (including transient errors), it records v1beta1.CheckoutFailedCondition=True and returns early.
+// On a successful checkout it removes v1beta1.CheckoutFailedCondition, and compares the current revision of HEAD to the
+// artifact on the object, and records v1beta1.ArtifactOutdatedCondition if they differ.
+// If instructed, the signature of the commit is verified if and recorded as v1beta1.SourceVerifiedCondition. If the
+// signature can not be verified or the verification fails, the Condition=False and it returns early.
+// If both the checkout and signature verification are successful, the given artifact pointer is set to a new artifact
+// with the available metadata.
+//
+// The caller should assume a failure if an error is returned, or the Result is zero.
+func (r *GitRepositoryReconciler) reconcileSource(ctx context.Context,
+	obj *sourcev1.GitRepository, artifact *sourcev1.Artifact, dir string) (ctrl.Result, error) {
+	// Configure authentication strategy to access the source
 	auth := &git.Auth{}
-	if repository.Spec.SecretRef != nil {
-		authStrategy, err := strategy.AuthSecretStrategyForURL(
-			repository.Spec.URL,
-			git.CheckoutOptions{
-				GitImplementation: repository.Spec.GitImplementation,
-				RecurseSubmodules: repository.Spec.RecurseSubmodules,
-			})
+	if obj.Spec.SecretRef != nil {
+		// Determine the auth strategy
+		authStrategy, err := strategy.AuthSecretStrategyForURL(obj.Spec.URL, git.CheckoutOptions{
+			GitImplementation: obj.Spec.GitImplementation,
+			RecurseSubmodules: obj.Spec.RecurseSubmodules,
+		})
 		if err != nil {
-			return sourcev1.GitRepositoryNotReady(repository, sourcev1.AuthenticationFailedReason, err.Error()), err
+			conditions.MarkTrue(obj, sourcev1.CheckoutFailedCondition, sourcev1.AuthenticationFailedReason,
+				"Failed to get auth strategy for Git implementation %q: %s", obj.Spec.GitImplementation, err)
+			// Do not return error as recovery without changes is impossible
+			return ctrl.Result{}, nil
```

Comment:
The issue I’m seeing with this approach (not returning errors) is that `flux logs` is no longer useful. You can’t determine what’s going on by tailing the logs anymore, as some critical errors only end up in conditions or/and events. While conditions make it easy for machines to determine state, the lack of logging hurts our end-users, as they have to look at logs, conditions and events to determine what’s going on, adding up to the cognitive load when debugging multiple controllers/resources at once. I guess from now on we should tell people to not really on logs, that means CloudWatch, Stackdriver, etc can’t be used for debugging, they can only determine state by connecting to Kubernetes API and run `flux get all` in a loop. This also means that platform admins must give devs vpn access to Kubernetes API, having access to a log aggregator UI is not sufficient to determine for example a miss configuration of Git auth.

Reply:
I am open to discuss options that add minimal logging so `flux logs` has value again, this may actually be combined with the debug suggestion I made in https://github.com/fluxcd/source-controller/pull/411#discussion_r680176785.

From an API conventions / Kubernetes observation perspective, logs generally should be your last resort, and you are normally expected to forward the Kubernetes Event to an aggregator for persistent storage.

Lets have a chat on Monday to see if we can agree on a solution we are both happy with.

Reply:
> From an API conventions / Kubernetes observation perspective, logs generally should be your last resort, and you are normally expected to forward the Kubernetes Event to an aggregator for persistent storage.

We can't impose this view of the world on our users, no managed Kubernetes offering does this, only logs are aggregated by default. Not to mention that not all Flux users can afford such an expensive feature, as most event  aggregators makes you pay at ingress, and Kubernetes only keeps the events for a short period of time in etcd, making the events unreliable for incident analysis.